### ISSUE TITLE
Support to get http response size by micrometer

### ISSUE THREAD

**Issue by dipindas** [micrometer-metrics/micrometer#880]
Please add support  to get http response size as a metric value

**Comment by jkschneider** [comment]
Thanks. Surprised we didn't think of that earlier. Will do it with a `DistributionSummary` so you can best understand the distribution of response sizes over time.

**Comment by mweirauch** [comment]
Just out of curiosity, would this include body-only-size or include the headers as well?

**Comment by shakuzen** [comment]
@mweirauch I think we should do both body and headers (with a tag?) as much as possible.  Also, a separate metric for request and response sizes would be potentially useful.
![both?](https://media.giphy.com/media/3o7aCRloybJlXpNjSU/giphy.gif)

**Comment by mweirauch** [comment]
Sounds good.
Would we be differentiating between server and client side instrumentation? Guess so.
So in analogy to `http.server.requests` and `http.client.requests`:
`http.server.request.size` and `http.client.request.size`? (Don't pin me on the naming.)

**Comment by izeye** [comment]
> `http.server.request.size` and `http.client.request.size`? (Don't pin me on the naming.)

@mweirauch I guess you meant `http.server.response.size` and `http.client.response.size`?

**Comment by fzyzcjy** [comment]
Hi is there any updates (after 2 years)?

**Comment by drriguz** [comment]
> Hi is there any updates (after 2 years)?

Hi is there any updates (after 3 years)?

**Comment by MayankManas1** [comment]
hi, is there any updates (after 4 years)?

**Comment by jonatan-ivanov** [comment]
There was:
>[shakuzen](https://github.com/shakuzen) modified the milestones: [1.x](https://github.com/micrometer-metrics/micrometer/milestone/37), [1.11 backlog](https://github.com/micrometer-metrics/micrometer/milestone/189) [on Jan 31](https://github.com/micrometer-metrics/micrometer/issues/880#event-8397950650)

It's not a guarantee that it will make it to 1.11 but issues like this (`help-wanted` label) are good for contributions is anyone wants to file a PR.

**Comment by babbarimam** [comment]
Hi @shakuzen ,
can you please confirm when and in which version we can expect the http response size metrics? 
Thanks

**Comment by bclozel** [comment]
From a Framework instrumentation perspective:

An easy way to instrument this would be to read the `Content-Length` HTTP header - unfortunately a lot of requests and responses use `Transfer-Encoding: chunked` and the only way to know the actual size is to wrap the request/responses and count the bytes as they're being read. Because this can happen with blocking and non-blocking I/O at the Servlet level, this can be quite challenging to implement and could cause a significant performance drop.

Also, even if this first problem is solved, the message size at the web Framework level is not the same as the actual message size on the network. Servers and Servlet containers in general support message compression and perform this task on-the-fly. This means that this metric instrumentation would need to happen at the server level, not at the web framework level. [This part is highlighted in the OTel semantic conventions as well](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpserverrequestsize) and the metric itself is optional.

To summarize, I think that mentioning `WebMvcMetricsFilter` is misleading here as the instrumentation must happen at the level of the HTTP library (server or client). For example, developers would need to use a binder to instrument Tomcat, not rely on the instrumentation done at the Spring Framework/Spring Boot level.

I think the best way to make progress on this is to review the available binders for HTTP servers and clients, and figure out if this type of instrumentation is possible in an efficient manner for each.

**Comment by kfko** [comment]
Hi is there any updates (after 7 years)?

**Comment by jonatan-ivanov** [comment]
Yes, see the comment before yours: https://github.com/micrometer-metrics/micrometer/issues/880#issuecomment-1625381050

**Comment by shakuzen** [comment]
While I understand this is a popular issue, I think the best way forward may be to close this and instead open issues for request/response size in specific HTTP server/client implementations where users want it. We don't have a generic HTTP instrumentation, and given the challenges highlighted in https://github.com/micrometer-metrics/micrometer/issues/880#issuecomment-1625381050, the details of the specific implementation are going to affect the instrumentation. Let's make this more actionable. I don't see any mention in any of the previous comments of a specific implementation. Only Spring's web framework is mentioned, which, as per https://github.com/micrometer-metrics/micrometer/issues/880#issuecomment-1625381050, is the wrong level to do this instrumentation at.

For what HTTP client/server specifically do people want this metric for?

**Comment by rupert-madden-abbott** [comment]
Personally, I am interested in having this for jetty.
